internal/controllers/topology/cluster/scope/state.go

```diff
@@ -123,3 +129,75 @@ func (md *MachineDeploymentState) IsUpgrading(ctx context.Context, c client.Clie
 	}
 	return false, nil
 }
+
+// MachinePoolsStateMap holds a collection of MachinePool states.
+type MachinePoolsStateMap map[string]*MachinePoolState
+
+// Upgrading returns the list of the machine pools
+// that are upgrading.
+func (mps MachinePoolsStateMap) Upgrading(ctx context.Context, c client.Client) ([]string, error) {
+	names := []string{}
+	for _, mp := range mps {
+		upgrading, err := mp.IsUpgrading(ctx, c)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to list upgrading MachinePools")
+		}
+		if upgrading {
+			names = append(names, mp.Object.Name)
+		}
+	}
+	return names, nil
+}
+
+// MachinePoolState holds all the objects representing the state of a managed pool.
+type MachinePoolState struct {
+	// Object holds the MachinePool object.
+	Object *expv1.MachinePool
+
+	// BootstrapObject holds the MachinePool bootstrap object.
+	BootstrapObject *unstructured.Unstructured
+
+	// InfrastructureMachinePoolObject holds the infrastructure machine template referenced by the MachinePool object.
+	InfrastructureMachinePoolObject *unstructured.Unstructured
+}
+
+// IsUpgrading determines if the MachinePool is upgrading.
+// A machine deployment is considered upgrading if at least one of the Machines of this
+// MachinePool has a different version.
+func (mp *MachinePoolState) IsUpgrading(ctx context.Context, c client.Client) (bool, error) {
+	// If the MachinePool has no version there is no definitive way to check if it is upgrading. Therefore, return false.
+	// Note: This case should not happen.
+	if mp.Object.Spec.Template.Spec.Version == nil {
+		return false, nil
+	}
+	machinePoolSelector := metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			clusterv1.MachinePoolNameLabel: format.MustFormatValue(mp.Object.Name),
+			clusterv1.ClusterNameLabel:     mp.Object.Spec.ClusterName,
+		},
+	}
+	selectorMap, err := metav1.LabelSelectorAsMap(&machinePoolSelector)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to check if MachinePool %s is upgrading: failed to convert label selector to map", mp.Object.Name)
+	}
+	machinePools := &expv1.MachinePoolList{}
+	if err := c.List(ctx, machinePools, client.InNamespace(mp.Object.Namespace), client.MatchingLabels(selectorMap)); err != nil {
+		return false, errors.Wrapf(err, "failed to check if MachinePool %s is upgrading: failed to list MachinePools", mp.Object.Name)
+	}
```

Comment:
```suggestion
```
We shouldn't need this. This func is only responsible to calculate if mp.Object is upgrading. No need to list MachinePools, we should just take mp.Object